File: TIA/tia_final_project/src/repositories/elitism/ElitismImplementation.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package repositories.elitism;

import configs.AppConfig;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import repositories.IndividualImplementation;
import repositories.IndividualInterface;

/**
 *
 * @author crdzbird
 */

public class ElitismImplementation implements ElitismInterface<IndividualImplementation> {

    public List<IndividualImplementation> elitism(List<IndividualImplementation> population, boolean compact) {
        List<IndividualImplementation> nextGeneration = new ArrayList<>(population.size());

        Collections.sort(population);

        int eliteCount = Math.round(population.size() * 0.2F);
        nextGeneration.addAll(population.subList(0, eliteCount));

        int startingCity = population.get(0).getGenotype().get(0);

        addNewCppIndividuals(nextGeneration, startingCity, eliteCount);

        while (nextGeneration.size() < population.size()) {
            IndividualImplementation parent1 = selectRandomIndividual(population);
            IndividualImplementation parent2 = selectRandomIndividual(population);
            nextGeneration.add((IndividualImplementation) parent1.getChildren((IndividualInterface) parent2).get(0));
        }

        if (compact) {
            Set<IndividualImplementation> compactedPop = new HashSet<>(nextGeneration);
            nextGeneration.clear();
            nextGeneration.addAll(compactedPop);
            addNewCppIndividuals(nextGeneration, startingCity, population.size() - nextGeneration.size());
        }

        Collections.sort(nextGeneration);

        return nextGeneration;
    }

    private void addNewCppIndividuals(List<IndividualImplementation> list, int startingCity, int count) {
        for (int i = 0; i < count; i++) {
            list.add(new IndividualImplementation(startingCity));
        }
    }

    private IndividualImplementation selectRandomIndividual(List<IndividualImplementation> population) {
        return population.get(AppConfig.RANDOM.nextInt(population.size()));
    }
}

File: TIA/tia_genetic_memetic/lib/repositories/elitism/elitism_implementation.dart
import 'dart:math';

import 'package:tia_genetic_memetic/repositories/elitism/elitism_abstraction.dart';
import 'package:tia_genetic_memetic/repositories/individual/individual_implementation.dart';

class ElitismImplementation
    implements ElitismInterface<IndividualImplementation> {
  @override
  List<IndividualImplementation> elitism(
      List<IndividualImplementation> population, bool compact) {
    List<IndividualImplementation> nextGeneration = List.from(population)
      ..sort();
    int eliteCount = (population.length * 0.2).round();
    nextGeneration = nextGeneration.sublist(0, eliteCount);
    int startingCity = population[0].getGenotype()[0];
    addNewCppIndividuals(nextGeneration, startingCity, eliteCount);
    while (nextGeneration.length < population.length) {
      IndividualImplementation parent1 = selectRandomIndividual(population);
      IndividualImplementation parent2 = selectRandomIndividual(population);
      nextGeneration
          .add(parent1.getChildren(parent2)[0] as IndividualImplementation);
    }
    if (compact) {
      Set<IndividualImplementation> compactedPop = Set.from(nextGeneration);
      nextGeneration.clear();
      nextGeneration.addAll(compactedPop);
      addNewCppIndividuals(nextGeneration, startingCity,
          population.length - nextGeneration.length);
    }
    nextGeneration.sort();
    return nextGeneration;
  }

  void addNewCppIndividuals(
      List<IndividualImplementation> list, int startingCity, int count) {
    for (int i = 0; i < count; i++) {
      list.add(IndividualImplementation(startingCity));
    }
  }

  IndividualImplementation selectRandomIndividual(
      List<IndividualImplementation> list) {
    Random random = Random();
    return list[random.nextInt(list.length)];
  }
}
